Hi, I've tried to test this with the [examples](https://github.com/zenaton/examples-python) repo, by commenting the first line `import client` of `boot.py` - but I've still have the error 
```
Traceback (most recent call last):
  File "/Users/gillesbarbier/.zenaton/lib/worker-0.7.6/priv/python/scripts/slave.py", line 35, in <module>
    slave.process()
  File "/Users/gillesbarbier/.zenaton/lib/worker-0.7.6/priv/python/lib/zenaton/loader/slave.py", line 25, in process
    return self.process_job(self.job)
  File "/Users/gillesbarbier/.zenaton/lib/worker-0.7.6/priv/python/lib/zenaton/loader/slave.py", line 30, in process_job
    zenaton.engine.Engine().processor = Processor()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/zenaton/singleton.py", line 6, in __call__
    cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/zenaton/engine.py", line 17, in __init__
    self.client = Client()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/zenaton/singleton.py", line 6, in __call__
    cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
TypeError: __init__() missing 3 required positional arguments: 'app_id', 'api_token', and 'app_env'
```
using this PR

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,8 @@
 ## [Unreleased]
 
 ### Added
+- No need for credentials when this lib is running in a Zenaton agent except if dispatching a
+  sub-job.
 
 ### Changed
 

--- a/tests/test_client.py
+++ b/tests/test_client.py
@@ -3,6 +3,7 @@ import pytest
 
 from zenaton.exceptions import InvalidArgumentError
 from zenaton.client import Client
+from zenaton.singleton import Singleton
 
 from .utils import validate_url
 
@@ -67,7 +68,29 @@ def test_event_workflow_lifecycle(client, sequential_workflow, my_event):
     assert response['status_code'] == 200
 
 
+@pytest.mark.usefixtures("client")
 def test_url_params_encoding(client):
     client.app_env = 'prod'
     assert client.add_app_env('', 'workflow_id').endswith('workflow_id')
     assert client.add_app_env('', 'yann+1@zenaton.com').endswith('&yann%2B1%40zenaton.com')
+
+
+def test_lazy_init():
+    Singleton._instances.clear()
+
+    # Start with a client without the API token.
+    client = Client()
+    # OK to contact the worker.
+    assert validate_url(client.worker_url())
+    # Not OK to contac the Zenaton API.
+    with pytest.raises(ValueError, match=r'.*API token.*'):
+        client.website_url()
+
+    # Create another client with the API token later.
+    client2 = Client('my-app-id', 'my-api-token', 'my-app-env')
+    # It's actually a singleton.
+    assert client == client2
+    assert validate_url(client2.worker_url())
+    website_url = client2.website_url()
+    assert validate_url(website_url)
+    assert 'my-api-token' in website_url

--- a/zenaton/client.py
+++ b/zenaton/client.py
@@ -42,13 +42,18 @@ class Client(metaclass=Singleton):
     WORKFLOW_PAUSE = 'pause'  # Worker udpate mode to pause a worker
     WORKFLOW_RUN = 'run'  # Worker update mode to resume a worker
 
-    def __init__(self, app_id, api_token, app_env):
+    def __init__(self, app_id='', api_token='', app_env=''):
         self.app_id = app_id
         self.api_token = api_token
         self.app_env = app_env
         self.http = HttpService()
         self.serializer = Serializer()
         self.properties = Properties()
+
+    def __lazy_init__(self, app_id, api_token, app_env):
+        self.app_id = self.app_id or app_id
+        self.api_token = self.api_token or api_token
+        self.app_env = self.app_env or app_env
 
     """
         Gets the url for the workers
@@ -70,6 +75,8 @@ class Client(metaclass=Singleton):
     """
     def website_url(self, resource='', params=''):
         api_url = os.environ.get('ZENATON_API_URL') or self.ZENATON_API_URL
+        if not self.api_token:
+            raise ValueError('Client not initialized to access website: missing an API token.')
         url = '{}/{}?{}={}&'.format(api_url, resource, self.API_TOKEN, self.api_token)
         return self.add_app_env(url, params)
 

--- a/zenaton/singleton.py
+++ b/zenaton/singleton.py
@@ -4,4 +4,6 @@ class Singleton(type):
     def __call__(cls, *args, **kwargs):
         if cls not in cls._instances:
             cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        elif (args or kwargs) and hasattr(cls._instances[cls], '__lazy_init__'):
+            cls._instances[cls].__lazy_init__(*args, **kwargs)
         return cls._instances[cls]

